ukrainian_word_stress/stressify_.py

```diff
@@ -0,0 +1,210 @@
+
+import marisa_trie
+import stanza
+
+
+log = logging.getLogger(__name__)
+
+
+class StressSymbol:
+    AcuteAccent = "´"
+    CombiningAcuteAccent = "\u0301"
+
+
+class OnAmbiguity:
+    Skip = "skip"
+    First = "first"
+    All = "all"
+
+
+class Stressifier:
+    """Add word stress to texts in Ukrainian.
+
+    Args:
+        `stress_symbol`: Which symbol to use as an accent mark.
+            Default is `StressSymbol.AcuteAccent` (я´йця)
+            Alternative is `StressSymbol.CombiningAcuteAccent` (я́йця).
+                This symbol is commonly used in print. However, not all
+                platforms render it correctly (Windows, for one).
+            Custom characters are also accepted.
+
+        `on_ambiguity`: What to do if word ambiguity cannot be resolved.
+            - `OnAmbiguity.Skip` (default): do not place stress
+            - `OnAmbiguity.First`: place a stress of the first match with a
+                high chance of being incorrect.
+            - `OnAmbiguity.All`: return all possible options at once.
+                This will look as multiple stress symbols in one word
+                (за´мо´к)
+
+    Example:
+        >>> f = Stressifier()
+        >>> f("Привіт, як справи?")
+        'Приві´т, як спра´ви?'
+    """
+
+
+
+    def __init__(self,
+                 stress_symbol=StressSymbol.AcuteAccent,
+                 on_ambiguity=OnAmbiguity.Skip):
+
+        dict_path = pkg_resources.resource_filename('ukrainian_word_stress', 'data/stress.v2.trie')
+        self.dict = marisa_trie.BytesTrie()
+        self.dict.load(dict_path)
+        self.nlp = stanza.Pipeline(
+            'uk',
+            processors='tokenize,pos,mwt',
+            download_method=stanza.pipeline.core.DownloadMethod.REUSE_RESOURCES,
+            logging_level=logging.getLevelName(log.getEffectiveLevel())
+        )
+        self.stress_symbol = stress_symbol
+        self.on_ambiguity = on_ambiguity
+
+    def __call__(self, text):
+        parsed = self.nlp(text)
+        result = MutableText(text)
+        log.debug("Parsed text: %s", parsed)
+        for token in parsed.iter_tokens():
+            accents = find_accent_positions(self.dict, token.to_dict()[0], self.on_ambiguity)
+            accented_token = self.apply_accent_positions(token.text, accents)
+            if accented_token != token:
+                result.replace(token.start_char, token.end_char, accented_token)
+
+        return result.get_edited_text()
+
+    def apply_accent_positions(self, s, positions):
+        for position in sorted(positions, reverse=True):
+            s = s[:position] + self.stress_symbol + s[position:]
+        return s
+
+
+def stressify(text: str) -> str:
+    """Add word stress to texts in Ukrainian.
+
+    This is a functional interface to `Stressifier`.
+
+    Example::
+        >>> stressify("Привіт, як справи?")
+        'Приві´т, як спра´ви?'
+
+    """
+
+    if not hasattr(stressify, "f"):
```

Comment:
Maybe something like this:
```python

class Singleton(type):
    _instances: dict = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super().__call__(*args, **kwargs)
        return cls._instances[cls]
```

Reply:
Thinking more about it, having a function that holds class instance(s) under the hood was not a good idea, considering that the class is quite heavy. To make it more transparent, I removed the function altogether.